Nested subgroups don't appear to work correctly

```python
def test_nested_subgroups():
    @dataclass
    class FooConfig:
        ...

    @dataclass
    class FooAConfig(FooConfig):
        foo_param_a: float = 0.0

    @dataclass
    class FooBConfig(FooConfig):
        foo_param_b: str = "foo_b"

    @dataclass
    class BarConfig:
        ...

    @dataclass
    class Bar1Config(BarConfig):
        bar_config_1: FooConfig = subgroups(
            {"foo_a": FooAConfig, "foo_b": FooBConfig},
            default=FooAConfig(),
        )

    @dataclass
    class Bar2Config(BarConfig):
        bar_config_2: FooConfig = subgroups(
            {"foo_a": FooAConfig, "foo_b": FooBConfig},
            default=FooBConfig(),
        )

    @dataclass
    class Config(TestSetup):
        bar_config: BarConfig = subgroups(
            {"bar_1": Bar1Config, "bar_2": Bar2Config},
            default=Bar2Config(),
        )

    assert Config.setup("") == Config(bar_config=Bar2Config(bar_config_2=FooBConfig()))
```

gives:
```console
>       assert Config.setup("") == Config(bar_config=Bar2Config(bar_config_2=FooBConfig()))

test/test_subgroups.py:319: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/testutils.py:174: in setup
    args = parser.parse_args(splits)
../../.conda/envs/simple/lib/python3.9/argparse.py:1826: in parse_args
    args, argv = self.parse_known_args(args, namespace)
simple_parsing/parsing.py:393: in parse_known_args
    parsed_args, unparsed_args = parser.parse_known_args(args, namespace=parsed_args)
simple_parsing/parsing.py:393: in parse_known_args
    parsed_args, unparsed_args = parser.parse_known_args(args, namespace=parsed_args)
simple_parsing/parsing.py:385: in parse_known_args
    parser.add_arguments(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = ArgumentParser(prog='pytest', usage=None, description=None, formatter_class=<class 'simple_parsing.help_formatter.SimpleHelpFormatter'>, conflict_handler='error', add_help=True)
dataclass = <class 'test.test_subgroups.test_nested_subgroups.<locals>.Bar2Config'>, dest = 'config.bar_config'

    def add_arguments(
        self,
        dataclass: type[Dataclass] | Dataclass,
        dest: str,
        *,
        prefix: str = "",
        default: Dataclass = None,
        dataclass_wrapper_class: type[DataclassWrapper] = DataclassWrapper,
    ):
        """Adds command-line arguments for the fields of `dataclass`.
    
        Parameters
        ----------
        dataclass : Union[Dataclass, Type[Dataclass]]
            The dataclass whose fields are to be parsed from the command-line.
            If an instance of a dataclass is given, it is used as the default
            value if none is provided.
        dest : str
            The destination attribute of the `argparse.Namespace` where the
            dataclass instance will be stored after calling `parse_args()`
        prefix : str, optional
            An optional prefix to add prepend to the names of the argparse
            arguments which will be generated for this dataclass.
            This can be useful when registering multiple distinct instances of
            the same dataclass, by default ""
        default : Dataclass, optional
            An instance of the dataclass type to get default values from, by
            default None
        """
        # Save this call so that we can replay it on any child later.
        self._add_argument_replay.append(
            partial(
                type(self).add_arguments,
                dataclass=dataclass,
                dest=dest,
                prefix=prefix,
                default=default,
                dataclass_wrapper_class=dataclass_wrapper_class,
            )
        )
        for wrapper in self._wrappers:
            if wrapper.dest == dest:
                if wrapper.dataclass == dataclass:
>                   raise argparse.ArgumentError(
                        argument=None,
                        message=f"Destination attribute {dest} is already used for "
                        f"dataclass of type {dataclass}. Make sure all destinations"
                        f" are unique. (new dataclass type: {dataclass})",
                    )
E                   argparse.ArgumentError: Destination attribute config.bar_config is already used for dataclass of type <class 'test.test_subgroups.test_nested_subgroups.<locals>.Bar2Config'>. Make sure all destinations are unique. (new dataclass type: <class 'test.test_subgroups.test_nested_subgroups.<locals>.Bar2Config'>)

simple_parsing/parsing.py:244: ArgumentError
====================================================================== short test summary info =======================================================================
FAILED test/test_subgroups.py::test_nested_subgroups - argparse.ArgumentError: Destination attribute config.bar_config is already used for dataclass of type <class...

```

--- a/examples/subgroups/subgroups_example.py
+++ b/examples/subgroups/subgroups_example.py
@@ -4,6 +4,7 @@ from dataclasses import dataclass
 from pathlib import Path
 
 from simple_parsing import ArgumentParser, subgroups
+from simple_parsing.wrappers.field_wrapper import ArgumentGenerationMode, NestedMode
 
 
 @dataclass
@@ -48,17 +49,19 @@ class Config:
     # Which model to use
     model: ModelConfig = subgroups(
         {"model_a": ModelAConfig, "model_b": ModelBConfig},
-        default=ModelAConfig(),
+        default_factory=ModelAConfig,
     )
 
     # Which dataset to use
     dataset: DatasetConfig = subgroups(
         {"dataset_1": Dataset1Config, "dataset_2": Dataset2Config},
-        default=Dataset2Config(),
+        default_factory=Dataset2Config,
     )
 
 
-parser = ArgumentParser()
+parser = ArgumentParser(
+    argument_generation_mode=ArgumentGenerationMode.NESTED, nested_mode=NestedMode.WITHOUT_ROOT
+)
 parser.add_arguments(Config, dest="config")
 args = parser.parse_args()
 

--- a/simple_parsing/parsing.py
+++ b/simple_parsing/parsing.py
@@ -14,7 +14,7 @@ from argparse import SUPPRESS, Action, HelpFormatter, Namespace, _
 from collections import defaultdict
 from logging import getLogger
 from pathlib import Path
-from typing import Any, Callable, Sequence, Type, TypeVar, overload
+from typing import Any, Callable, Sequence, Type, overload
 
 from simple_parsing.helpers.subgroups import SubgroupKey
 from simple_parsing.wrappers.dataclass_wrapper import DataclassWrapperType
@@ -971,14 +971,12 @@ class ArgumentParser(argparse.ArgumentParser):
         return leftover_args, constructor_arguments
 
 
-T = TypeVar("T")
-
-
+# TODO: Change the order of arguments to put `args` as the second argument.
 def parse(
-    config_class: type[Dataclass],
+    config_class: type[DataclassT],
     config_path: Path | str | None = None,
     args: str | Sequence[str] | None = None,
-    default: Dataclass | None = None,
+    default: DataclassT | None = None,
     dest: str = "config",
     *,
     prefix: str = "",
@@ -989,7 +987,7 @@ def parse(
     formatter_class: type[HelpFormatter] = SimpleHelpFormatter,
     add_config_path_arg: bool | None = None,
     **kwargs,
-) -> Dataclass:
+) -> DataclassT:
     """Parse the given dataclass from the command-line.
 
     See the `ArgumentParser` constructor for more details on the arguments (they are the same here

--- a/test/test_subgroups.py
+++ b/test/test_subgroups.py
@@ -14,7 +14,8 @@ import pytest
 from pytest_regressions.file_regression import FileRegressionFixture
 from typing_extensions import Annotated
 
-from simple_parsing import ArgumentParser, subgroups
+from simple_parsing import ArgumentParser, parse, subgroups
+from simple_parsing.wrappers.field_wrapper import ArgumentGenerationMode, NestedMode
 
 from .test_choice import Color
 from .testutils import TestSetup, raises_invalid_choice, raises_missing_required_arg
@@ -839,3 +840,94 @@ def test_nested_subgroups(frozen: bool):
 
     assert Config.setup("") == Config(bar=Bar2Config(foo=FooBConfig()))
     assert Config.setup("--bar=bar_1 --foo=foo_a") == Config(bar=Bar1Config(foo=FooAConfig()))
+
+
+@dataclass
+class ModelConfig:
+    ...
+
+
+@dataclass
+class DatasetConfig:
+    ...
+
+
+@dataclass
+class ModelAConfig(ModelConfig):
+    lr: float = 3e-4
+    optimizer: str = "Adam"
+    betas: tuple[float, float] = (0.9, 0.999)
+
+
+@dataclass
+class ModelBConfig(ModelConfig):
+    lr: float = 1e-3
+    optimizer: str = "SGD"
+    momentum: float = 1.234
+
+
+@dataclass
+class Dataset1Config(DatasetConfig):
+    data_dir: str | Path = "data/foo"
+    foo: bool = False
+
+
+@dataclass
+class Dataset2Config(DatasetConfig):
+    data_dir: str | Path = "data/bar"
+    bar: float = 1.2
+
+
+@dataclass
+class Config(TestSetup):
+
+    # Which model to use
+    model: ModelConfig = subgroups(
+        {"model_a": ModelAConfig, "model_b": ModelBConfig},
+        default_factory=ModelAConfig,
+    )
+
+    # Which dataset to use
+    dataset: DatasetConfig = subgroups(
+        {"dataset_1": Dataset1Config, "dataset_2": Dataset2Config},
+        default_factory=Dataset2Config,
+    )
+
+
+def _parse_config(args: str) -> Config:
+    return parse(
+        Config,
+        args=args,
+        argument_generation_mode=ArgumentGenerationMode.NESTED,
+        nested_mode=NestedMode.WITHOUT_ROOT,
+    )
+
+
+def test_ordering_of_args_doesnt_matter():
+    """Test to confirm that #160 is fixed:"""
+
+    # $ python issue.py --model model_a --model.lr 1e-2
+    assert _parse_config(args="--model model_a --model.lr 1e-2") == Config(
+        model=ModelAConfig(lr=0.01, optimizer="Adam", betas=(0.9, 0.999)),
+        dataset=Dataset2Config(data_dir="data/bar", bar=1.2),
+    )
+
+    # I was expecting this to work given that both model configs have `lr` attribute
+    # $ python issue.py --model.lr 1e-2.
+    assert _parse_config(args="--model.lr 1e-2") == Config(
+        model=ModelAConfig(lr=1e-2, optimizer="Adam", betas=(0.9, 0.999)),
+        dataset=Dataset2Config(data_dir="data/bar", bar=1.2),
+    )
+
+    # $ python issue.py --model model_a --model.betas 0. 1.
+    assert _parse_config(args="--model model_a --model.betas 0. 1.") == Config(
+        model=ModelAConfig(lr=0.0003, optimizer="Adam", betas=(0.0, 1.0)),
+        dataset=Dataset2Config(data_dir="data/bar", bar=1.2),
+    )
+
+    # % ModelA being the default, I was expecting this two work
+    # $ python issue.py --model.betas 0. 1.
+    assert _parse_config(args="--model.betas 0. 1.") == Config(
+        model=ModelAConfig(lr=0.0003, optimizer="Adam", betas=(0.0, 1.0)),
+        dataset=Dataset2Config(data_dir="data/bar", bar=1.2),
+    )

--- a/test/test_subgroups.py
+++ b/test/test_subgroups.py
@@ -794,3 +794,48 @@ We expect to get:
 #     assert Config.setup("--model small").model == SmallModel()
 #     assert Config.setup("--model big").model == BigModel()
 #     assert Config.setup("--num_layers 123").model == Model(num_layers=123, hidden_dim=32)
+
+
+@pytest.mark.parametrize("frozen", [True, False])
+def test_nested_subgroups(frozen: bool):
+    """Assert that #160 is fixed: https://github.com/lebrice/SimpleParsing/issues/160"""
+
+    @dataclass(frozen=frozen)
+    class FooConfig:
+        ...
+
+    @dataclass(frozen=frozen)
+    class BarConfig:
+        foo: FooConfig
+
+    @dataclass(frozen=frozen)
+    class FooAConfig(FooConfig):
+        foo_param_a: float = 0.0
+
+    @dataclass(frozen=frozen)
+    class FooBConfig(FooConfig):
+        foo_param_b: str = "foo_b"
+
+    @dataclass(frozen=frozen)
+    class Bar1Config(BarConfig):
+        foo: FooConfig = subgroups(
+            {"foo_a": FooAConfig, "foo_b": FooBConfig},
+            default_factory=FooAConfig,
+        )
+
+    @dataclass(frozen=frozen)
+    class Bar2Config(BarConfig):
+        foo: FooConfig = subgroups(
+            {"foo_a": FooAConfig, "foo_b": FooBConfig},
+            default_factory=FooBConfig,
+        )
+
+    @dataclass(frozen=frozen)
+    class Config(TestSetup):
+        bar: Bar1Config | Bar2Config = subgroups(
+            {"bar_1": Bar1Config, "bar_2": Bar2Config},
+            default_factory=Bar2Config,
+        )
+
+    assert Config.setup("") == Config(bar=Bar2Config(foo=FooBConfig()))
+    assert Config.setup("--bar=bar_1 --foo=foo_a") == Config(bar=Bar1Config(foo=FooAConfig()))

--- a/test/test_subgroups/test_help[Config---help].md
+++ b/test/test_subgroups/test_help[Config---help].md
@@ -1,4 +1,4 @@
-# Regression file for [this test](test/test_subgroups.py:724)
+# Regression file for [this test](test/test_subgroups.py:725)
 
 Given Source code:
 

--- a/test/test_subgroups/test_help[Config---model=model_a --help].md
+++ b/test/test_subgroups/test_help[Config---model=model_a --help].md
@@ -1,4 +1,4 @@
-# Regression file for [this test](test/test_subgroups.py:724)
+# Regression file for [this test](test/test_subgroups.py:725)
 
 Given Source code:
 

--- a/test/test_subgroups/test_help[Config---model=model_b --help].md
+++ b/test/test_subgroups/test_help[Config---model=model_b --help].md
@@ -1,4 +1,4 @@
-# Regression file for [this test](test/test_subgroups.py:724)
+# Regression file for [this test](test/test_subgroups.py:725)
 
 Given Source code:
 

--- a/test/test_subgroups/test_help[ConfigWithFrozen---conf=even --a 100 --help].md
+++ b/test/test_subgroups/test_help[ConfigWithFrozen---conf=even --a 100 --help].md
@@ -1,4 +1,4 @@
-# Regression file for [this test](test/test_subgroups.py:724)
+# Regression file for [this test](test/test_subgroups.py:725)
 
 Given Source code:
 

--- a/test/test_subgroups/test_help[ConfigWithFrozen---conf=even --help].md
+++ b/test/test_subgroups/test_help[ConfigWithFrozen---conf=even --help].md
@@ -1,4 +1,4 @@
-# Regression file for [this test](test/test_subgroups.py:724)
+# Regression file for [this test](test/test_subgroups.py:725)
 
 Given Source code:
 

--- a/test/test_subgroups/test_help[ConfigWithFrozen---conf=odd --a 123 --help].md
+++ b/test/test_subgroups/test_help[ConfigWithFrozen---conf=odd --a 123 --help].md
@@ -1,4 +1,4 @@
-# Regression file for [this test](test/test_subgroups.py:724)
+# Regression file for [this test](test/test_subgroups.py:725)
 
 Given Source code:
 

--- a/test/test_subgroups/test_help[ConfigWithFrozen---conf=odd --help].md
+++ b/test/test_subgroups/test_help[ConfigWithFrozen---conf=odd --help].md
@@ -1,4 +1,4 @@
-# Regression file for [this test](test/test_subgroups.py:724)
+# Regression file for [this test](test/test_subgroups.py:725)
 
 Given Source code:
 

--- a/test/test_subgroups/test_help[ConfigWithFrozen---help].md
+++ b/test/test_subgroups/test_help[ConfigWithFrozen---help].md
@@ -1,4 +1,4 @@
-# Regression file for [this test](test/test_subgroups.py:724)
+# Regression file for [this test](test/test_subgroups.py:725)
 
 Given Source code:
 

